### ISSUE TITLE
Understanding Fisher et al. 2019: All Models are Wrong, but Many are Useful 

### ISSUE THREAD

**Issue by KarolinaChalupova** [KarolinaChalupova/DiplomaThesis#8]
I think this paper could be crucial. https://arxiv.org/abs/1801.01489 
With a single measure: 
- It gives confidence intervals for feature importance
- helps understand why ensembles work
- helps with the problem of interpreting models with correlated features

If I understand correctly, the MR is permutation feature importance (based on decrease in loss). MCR provides confidence interval for MR, using epsilon-Rashomon set. 
     - Issue A: I cannot find an implementation in python. I can try to code it up. There is R interpretation from the authors: https://github.com/aaronjfisher/mcr
     - Issue B: I do not understand what models consitute the empirical epsilon-Rashomon set - what is their general family - different models / architectures / seeds?

**Comment by KarolinaChalupova** [comment]
@martinhronec @barunik  I think this is a very interesting paper. However, I do not see how the empirical epsilon-Rashomon set should be constructed (see above).  If you think it is worth looking into, could you give me a hint on that please?

**Comment by KarolinaChalupova** [comment]
Update: I implemented ModelReliance and ModelClassReliance it in file interpretation.py, ModelReliance works and I am quite sure I have it right, ModelClassReliance still untested and I'm not so sure I got it right.

**Comment by barunik** [comment]
This is really interesting work, and I think you understand it correctly. Now to the issues: It seems you already solved A (imoplementation), and as for B, if I understand the paper correctly, this all works for linear models only, but for the models where standar epiral loss minimization is not known as with neural networks, or other complex nonlinear problems, MCR will not help. To me it seems that a general family for empirical epsilon-Rashomon set is a family of different linear combinations of variables that may fit the data equally well. But with ML models, this is not feasible? I need to read more deeply maybe.

**Comment by KarolinaChalupova** [comment]
@barunik Yes, that is my understanding as well. There should be no problem in using MR, but I don't see how MCR could work. Instead, to get the confidence intervals, I could maybe use bootstrap (calculate MR on randomly drawn observations a number of times)? Would it make sense?

**Comment by barunik** [comment]
Maybe I am bit confused how you like to us this. My answer was to initial issue B about family of models. Could you be bit more specific on how you like to use MR/MCR in your work so I can answer?

**Comment by KarolinaChalupova** [comment]
It is just an idea, but I was thinking of using MR and MCR as feature importance measures as follows. 
1) MR - for given model, calculate importance of individual features - (given model means single depth, single train-test split, single seed or ensemble -> single number for each feature)
2) Give confidence intervals for MR for given model, to be able to say if a feature is statistically important or not. I thought of doing this either by MCR (2A) or by bootstrap (2B):  
2A) Calculate MCR of a model class family. Now the question is how to define the model class. I was thinking about aggregation across random seeds and/or across model architectures.  (Aggregating across seeds is in my case impossible for ensembles, as the seed is already aggregated to create the ensemble.) 
2B) Calculate multiple MR for given model (single depth, single train-test split, single seed or ensemble) by bootstraping the set on which the measure is calculated.

**Comment by barunik** [comment]
It is a good idea indeed! And really nice paper (I have it on the reading list, especially theory part) by the way. But if I understand it correctly, it all works only for linear models, that is why I was confused since your models for both 1. and 2. (that is MR for a given model or MCR for a class of models) are highly nonlinear and I am not sure how these results hold for such nonlinear models. Authors state in the paper that it is an open problem and we do not know how to compute these stastitsics for nonlinear complaex models. That is what I meant in earlier comment.

Of course we can have a short meeting instead to clarify this so I do not slow you down with my confusion.

**Comment by KarolinaChalupova** [comment]
@barunik 
I have checked how the interpretation of the same model changes across different random seeds and got some very interesting results. It really helps to answer the question how much confidence we can put into the interpretation of which model. There are plenty of interesting results I can calculate based on this idea, so I think they may be no need to dive into MCR, even though it is interesting.  I will shortly list the preliminary results in the README. As for MR, I think it is safe to use it as one of the measures of feature importance, since MR as opposed to MCR has nothing to do with confidence, so using it simply as a measure should be fine (Molnar's textbook lists it as a measure for example). In the meantime, if you agree, I think I can close this issue end we can return to it outside the scope of the thesis.

**Comment by KarolinaChalupova** [comment]
@barunik @martinhronec I can show you the results for the different random seeds soon, we can have a call, I think they are pretty great. I will let you know as soon as I have it ready in a representable form, in couple of days tops. In the meantime, I will discuss the results briefly in README to give you an idea. There is plenty of room to build on the results too.

**Comment by barunik** [comment]
...looking forward to review the results!
